pkg/local_object_storage/engine/put.go

```diff
@@ -76,38 +79,77 @@ func (e *StorageEngine) put(prm PutPrm) (PutRes, error) {
 	}
 
 	finished := false
+	var bestShard hashedShard
+	var bestPool util.WorkerPool
 
 	e.iterateOverSortedShards(addr, func(ind int, sh hashedShard) (stop bool) {
 		e.mtx.RLock()
 		pool, ok := e.shardPools[sh.ID().String()]
+		if ok && bestPool == nil {
+			bestShard = sh
+			bestPool = pool
+		}
 		e.mtx.RUnlock()
 		if !ok {
 			// Shard was concurrently removed, skip.
 			return false
 		}
 
-		putDone, exists := e.putToShard(sh, ind, pool, addr, prm)
-		finished = putDone || exists
+		exists, err := e.putToShard(sh, ind, pool, addr, prm)
+		finished = err == nil || exists
 		return finished
 	})
 
 	if !finished {
-		err = errPutShard
+		err = e.putToShardWithDeadLine(bestShard, 0, bestPool, addr, prm)
+		if err != nil {
+			e.log.Warn("last stand to put object to the best shard",
+				zap.Stringer("addr", addr),
+				zap.Stringer("shard", bestShard.ID()),
+				zap.Error(err))
+
+			return PutRes{}, errPutShard
+		}
 	}
 
-	return PutRes{}, err
+	return PutRes{}, nil
+}
+
+func (e *StorageEngine) putToShardWithDeadLine(sh hashedShard, ind int, pool util.WorkerPool, addr oid.Address, prm PutPrm) error {
+	const deadline = 30 * time.Second
+	timer := time.NewTimer(deadline)
+	defer timer.Stop()
+
+	const putCooldown = 100 * time.Millisecond
+	ticker := time.NewTicker(putCooldown)
```

Comment:
One of the reasons I don't like pools we have now. Channels provide natural queueing and blocking/unblocking.